src/http/content.cr

```diff
@@ -47,106 +49,142 @@ module HTTP
 
-      to_read = Math.min(slice.size, @chunk_remaining)
+      to_read = Math.min(count, @chunk_remaining)
 
       bytes_read = @io.read slice[0, to_read]
-      @chunk_remaining -= bytes_read
 
-      check_chunk_remaining_is_zero
+      if bytes_read == 0
+        raise IO::EOFError.new("Invalid HTTP chunked content: unexpected end of file")
+      end
+
+      chunk_bytes_read bytes_read
 
       bytes_read
     end
 
     def read_byte
-      if @chunk_remaining > 0
-        byte = @io.read_byte
-        if byte
-          @chunk_remaining -= 1
-          check_chunk_remaining_is_zero
-        end
+      next_chunk
+      return super if @chunk_remaining == 0
+
+      byte = @io.read_byte
+      if byte
+        chunk_bytes_read 1
         byte
       else
-        super
+        raise IO::EOFError.new("Invalid HTTP chunked content: unexpected end of file")
       end
     end
 
     def peek
-      while true
-        if @chunk_remaining > 0
-          peek = @io.peek
-          return nil unless peek
-
-          if @chunk_remaining < peek.size
-            peek = peek[0, @chunk_remaining]
-          end
+      next_chunk
+      return nil if @chunk_remaining == 0
 
-          return peek
-        elsif @read_chunk_start
-          read_chunk_start
-          next
-        end
+      peek = @io.peek || return
 
-        break
+      if @chunk_remaining < peek.size
+        peek = peek[0, @chunk_remaining]
+      elsif peek.size == 0
+        raise IO::EOFError.new("Invalid HTTP chunked content: unexpected end of file")
       end
 
-      nil
+      peek
     end
 
     def skip(bytes_count)
       if bytes_count <= @chunk_remaining
         @io.skip(bytes_count)
-        @chunk_remaining -= bytes_count
-        check_chunk_remaining_is_zero
+        chunk_bytes_read bytes_count
       else
         super
       end
     end
 
-    private def check_chunk_remaining_is_zero
+    private def chunk_bytes_read(size)
+      @chunk_remaining -= size
+
       # As soon as we finish reading a chunk we return,
       # in case the next content is delayed (see #3270).
-      # We set @read_chunk_start to true so we read the next
+      # We set @expect_chunk_start to true so we read the next
       # chunk start on the next call to `read`.
       if @chunk_remaining == 0
-        @read_chunk_start = true
+        read_crlf
+        @expect_chunk_start = true
       end
     end
 
-    private def read_chunk_start
-      read_chunk_end
-      @chunk_remaining = @io.gets.not_nil!.to_i(16)
-      check_last_chunk
-      @read_chunk_start = false
+    # Check if the last read consumed a chunk and we
+    # need to start consuming the next one.
+    private def next_chunk
+      return unless @expect_chunk_start
+
+      if read_chunk_size == 0
+        read_trailer
+      end
+
+      @expect_chunk_start = false
     end
 
-    private def read_chunk_end
-      # Read "\r\n"
-      @io.skip(2)
+    private def read_crlf
+      bytes = Bytes.new(2)
```

Comment:
```crystal
bytes = uninitialized UInt8[2]
```

Otherwise 2 bytes are allocated on the HEAP memory each time the method is called :fearful: